### PR TITLE
download all necessary MeSH files + improve MeSHDownloader interface

### PR DESCRIPTION
**Pull request by fenekku**
Part of closing https://github.com/galterlibrary/Prism/issues/641 .

- download all necessary MeSH files
- improve MeSHDownloader interface
- set up CLI + tests for upcoming upgrade work

### CODE REVIEW

File: galter_subjects_utils/cli.py [hunk 1]
```diff
@@ -52,36 +47,71 @@ def to_mesh_downloader_kwargs(parameters):
...
-@click.option(
-    "--filter", "-f",
-    type=click.Choice(["topic", "topic-qualifier"]),
-    default=defaults["filter"],
-)
-@click.option(
-    "--downloads-dir", "-d",
-    type=click.Path(path_type=Path),
-    default=defaults["downloads-dir"])
+def mesh_download_options(f):
```
Comment (Meowcenary) [inline galter_subjects_utils/cli.py:62]: Nice! 👍 

It's interesting to see someone skillfully use decorators. Definitely one one of the Python features that I struggle with.

File: tests/test_mesh.py [hunk 2]
```diff
@@ -51,25 +44,31 @@ def assert_includes(dicts, dict_cores):
...
+        year="2023",
         directory=downloads_dir
     )
-
     downloader.download()
 
     patched_get.assert_called()
-    assert downloads_dir / "d2022.bin" == downloader.prefix_to_filepath["d"]
-    assert downloads_dir / "q2022.bin" == downloader.prefix_to_filepath["q"]
+    assert (downloads_dir / "d2023.bin").exists()
```
Comment (Meowcenary) [inline tests/test_mesh.py:64]: Ha, took me a second to realize these are **2023** which made it clear what was going on 😆

File: galter_subjects_utils/downloader.py [hunk 3]
```diff
@@ -26,30 +27,82 @@ def download_file(url, dest):
...
+    def file_urls(self):
+        """List of files to download."""
+        current_year = date.today().year
+        url_parent_segment = "MESH_FILES" if self.year == current_year else self.year  # noqa
+        base_url = self.base_url + "/" + url_parent_segment
+
+        urls = []
+
+        # asciimesh files
+        base_url_asciimesh = base_url + "/" + "asciimesh"
```
Comment (Meowcenary) [inline galter_subjects_utils/downloader.py:52]: It's almost tempting to wrap these blocks into even smaller functions, but that would probably just be over engineered at that point.

**Review by Meowcenary — APPROVED** [review]
Looks good to me! I think I was able to follow it pretty well, but as usual my comments are more just thoughts I had while reviewing.

File: galter_subjects_utils/downloader.py (hunk 3 above)
Comment (fenekku) [reply galter_subjects_utils/downloader.py:52]: Tell me about it :face_exhaling: : I originally lost a good number of hours seeing if a dynamic `getattr()` could be used to DRY-out the 4 URL construction and filepaths getters, but that way lied madness.

**Review by mattbaumann1 — APPROVED** [review]
Looks good to me. My comment is more of a FYI than anything that needs to be addressed.

File: galter_subjects_utils/downloader.py (hunk 3 above)
Comment (fenekku) [reply galter_subjects_utils/downloader.py:52]: Yup that's exactly the `getattr()` mentioned above... What becomes abstracted for what potential future and at what cost became too much to factor for me. I find myself gravitating towards a more and more static style personally for simplicity.

This is definitely the kind of thing that binds me in knots to be honest and I try to cut through it by keeping the end goal in mind. But it's a constant struggle as mentioned.

Comment (mattbaumann1) [reply galter_subjects_utils/downloader.py:52]: @fenekku, I think that's fair. I was trying to remember, but I think I saw something about how costly define_method in Rails is. I'm sorry I can't remember the details. Anyway, all that makes sense.

**Pull request merged**